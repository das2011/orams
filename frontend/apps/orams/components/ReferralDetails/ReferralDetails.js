import React, { Component } from 'react'
import parseDate from 'date-fns/parse'
import format from 'date-fns/format'
import AUpageAlert from '@gov.au/page-alerts/lib/js/react.js'
import LoadingIndicatorFullPage from 'shared/LoadingIndicatorFullPage/LoadingIndicatorFullPage'
import styles from './ReferralDetails.scss'

const DETAILS_KEY_MAP = {
  supplier: 'Supplier',
  price: 'Price',
  dateCreated: 'Date of creation',
  location: 'Location',
  createdBy: 'Created By',
  referralId: 'Referral ID',
  agency: 'Agency'
}

export default class ReferralDetails extends Component {
  static renderDetailsItem(title, value) {
    return (
      <div className={`row ${styles.detailsRow}`} key={title}>
        <div className="col-sm-3 col-xs-12">
          <div className={styles.title}>
            {title}
          </div>
        </div>
        <div>
          {value}
        </div>
      </div>
    )
  }

  static renderDisclaimerText() {
    return <div className="row">Disclaimer:</div>
  }

  static renderHorizontalLine() {
    return <div className={styles.horizontalSeparatorLine} />
  }

  renderAllDetails() {
    const details = { ...this.props.referralDetails }
    const referralDetails = {
      supplier: details.supplier,
      price: `$${details.price} incl GST`,
      dateCreated: format(parseDate(details.dateCreated), 'DD MMM YYYY'),
      location: `${details.regionName} ${details.regionState}`,
      createdBy: details.createdBy,
      referralId: details.referralId,
      agency: details.agency
    }

    return Object.keys(DETAILS_KEY_MAP).map(k =>
      ReferralDetails.renderDetailsItem(DETAILS_KEY_MAP[k], referralDetails[k])
    )
  }

  render() {
    if (this.props.isLoading) {
      return <LoadingIndicatorFullPage />
    } else if (this.props.errorMessage) {
      return (
        <div className={styles.banner}>
          <AUpageAlert as="error">
            <h4>
              {this.props.errorMessage}
            </h4>
          </AUpageAlert>
        </div>
      )
    }
    return (
      <main>
        <div className={`row ${styles.headerSection}`}>
          <div className="col-xs-12 col-sm-12">
            <h1 className="au-display-xl">Referral Details</h1>
          </div>
        </div>

        {this.renderAllDetails()}

        {ReferralDetails.renderHorizontalLine()}
        {ReferralDetails.renderDisclaimerText()}
      </main>
    )
  }
}
