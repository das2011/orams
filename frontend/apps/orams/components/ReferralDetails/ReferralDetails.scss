.banner {
  margin-top: 114px;
  margin-bottom: 114px;
}

.headerSection {
  margin-bottom: 1.5rem;
}

.title {
  font-size: 18px;
  font-weight: bold;
  color: #202751;
}

.detailsRow {
  margin-top: 12px;
  margin-bottom: 12px;
}

.horizontalSeparatorLine {
  border-top: solid 1px #bbbbbb;
  margin-top: 48px;
  margin-bottom: 48px;
}