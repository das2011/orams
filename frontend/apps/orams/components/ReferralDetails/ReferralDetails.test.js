import React from 'react'
import snapshotRenderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import ReferralDetails from './ReferralDetails'

Enzyme.configure({ adapter: new Adapter() })

describe('ReferralDetails component', () => {
  const defaultProps = {
    referralDetails: {
      supplier: 'Supplier name',
      price: 123,
      dateCreated: '2018-06-07T12:13:24.597Z',
      regionName: 'East',
      regionState: 'NSW',
      createdBy: 'Stan Lee',
      referralId: '456',
      agency: 'ATO'
    }
  }

  it('renders correctly when error message is present', () => {
    const props = {
      errorMessage: 'This is an error'
    }
    const tree = snapshotRenderer.create(<ReferralDetails {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when page is loading', () => {
    const props = {
      isLoading: true
    }
    const tree = snapshotRenderer.create(<ReferralDetails {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when there is no error', () => {
    const tree = snapshotRenderer.create(<ReferralDetails {...defaultProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with button for supplier', () => {
    const props = {
      ...defaultProps,
      userType: 'supplier'
    }
    const tree = snapshotRenderer.create(<ReferralDetails {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handles Accept referral click by supplier', () => {
    const mockAcceptReferral = jest.fn()
    const mockHistoryPush = jest.fn()
    const props = {
      ...defaultProps,
      userType: 'supplier',
      acceptReferral: mockAcceptReferral,
      history: {
        push: mockHistoryPush
      }
    }
    const component = shallow(<ReferralDetails {...props} />)

    component.find('.referral--accept').simulate('click')

    expect(mockAcceptReferral).toHaveBeenCalledTimes(1)
    expect(mockAcceptReferral).toHaveBeenCalledWith('456')
    expect(mockHistoryPush).toHaveBeenCalledTimes(1)
    expect(mockHistoryPush).toHaveBeenCalledWith('/dashboard')
  })
})
