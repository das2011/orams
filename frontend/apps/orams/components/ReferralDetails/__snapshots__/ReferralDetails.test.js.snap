// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferralDetails component renders correctly when error message is present 1`] = `
<div
  className="banner"
>
  <div
    className="au-body au-page-alerts   au-page-alerts--error"
    role="alert"
  >
    <h4>
      This is an error
    </h4>
  </div>
</div>
`;

exports[`ReferralDetails component renders correctly when page is loading 1`] = `
<div
  className="wrapper"
>
  <div
    className="loadingIndicator"
  >
    <div
      className="loadingInner"
    />
  </div>
  <div
    className="loadingText"
  >
    Loading...
  </div>
</div>
`;

exports[`ReferralDetails component renders correctly when there is no error 1`] = `
<main>
  <div
    className="row headerSection"
  >
    <div
      className="col-xs-12 col-sm-12"
    >
      <h1
        className="au-display-xl"
      >
        Referral Details
      </h1>
    </div>
  </div>
  <div
    className="row detailsRow"
  >
    <div
      className="col-sm-3 col-xs-12"
    >
      <div
        className="title"
      >
        Supplier
      </div>
    </div>
    <div>
      Supplier name
    </div>
  </div>
  <div
    className="row detailsRow"
  >
    <div
      className="col-sm-3 col-xs-12"
    >
      <div
        className="title"
      >
        Price
      </div>
    </div>
    <div>
      $123 incl GST
    </div>
  </div>
  <div
    className="row detailsRow"
  >
    <div
      className="col-sm-3 col-xs-12"
    >
      <div
        className="title"
      >
        Date of creation
      </div>
    </div>
    <div>
      7 Jun 2018
    </div>
  </div>
  <div
    className="row detailsRow"
  >
    <div
      className="col-sm-3 col-xs-12"
    >
      <div
        className="title"
      >
        Location
      </div>
    </div>
    <div>
      NSW East
    </div>
  </div>
  <div
    className="row detailsRow"
  >
    <div
      className="col-sm-3 col-xs-12"
    >
      <div
        className="title"
      >
        Created By
      </div>
    </div>
    <div>
      Stan Lee
    </div>
  </div>
  <div
    className="row detailsRow"
  >
    <div
      className="col-sm-3 col-xs-12"
    >
      <div
        className="title"
      >
        Referral ID
      </div>
    </div>
    <div>
      456
    </div>
  </div>
  <div
    className="row detailsRow"
  >
    <div
      className="col-sm-3 col-xs-12"
    >
      <div
        className="title"
      >
        Agency
      </div>
    </div>
    <div>
      ATO
    </div>
  </div>
  <div
    className="horizontalSeparatorLine"
  />
  <div
    className="row"
  >
    Disclaimer:
  </div>
</main>
`;
